lib/ancestry/instance_methods.rb

```diff
@@ -151,17 +155,45 @@ def ancestor_of?(node)
     # assuming that parent hasn't changed
     def parent= parent
       self.ancestor_ids = parent ? parent.path_ids : []
+      clear_memoized_instance_variables
     end
 
     def parent_id= new_parent_id
       self.parent = new_parent_id.present? ? unscoped_find(new_parent_id) : nil
+      clear_memoized_instance_variables
     end
 
     def parent_id
-      ancestor_ids.last if ancestors?
+      return @_parent_id if defined?(@_parent_id)
+      @_parent_id = if @_ancestor_ids
+        @_ancestor_ids.empty? ? nil : @_ancestor_ids.last
+      else
+        col = read_attribute(ancestry_base_class.ancestry_column)
+        # Specifically not using `.blank?` here because it is
+        # slower than doing the below.
+        if col.nil? || col.empty? # rubocop:disable Rails/Blank
+          nil
+        else
+          rindex = col.rindex(ANCESTRY_DELIMITER)
+          cast_primary_key(rindex ? col[rindex + 1, col.length] : col)
+        end
+      end
     end
     alias :parent_id? :ancestors?
 
+    def depth
+      @_depth ||= if @_ancestor_ids
+        @_ancestor_ids.size
+      else
+        col = read_attribute(ancestry_base_class.ancestry_column)
+        col ? col.count(ANCESTRY_DELIMITER) + 1 : 0
+      end
+    end
+
+    def cast_primary_key(key)
```

Comment:
I could have sworn that this was no longer necessary.

aah - it is still used in parent_id - if we drop that optimization, then we drop this method